### ISSUE TITLE
Bandwidth of a MIDI device

### ISSUE THREAD

**Issue by vincentbernat** [waltervn/opl3_fpga-apps#2]
Hey!

By curiosity, do you know what the bandwidth of a MIDI device is? Did you try to play Another World? They use the OPL to play digitized sound as either 5kHz or 10kHz.

**Comment by waltervn** [comment]
"Hardware" MIDI speed is 31250 bits per second, with one start and stop bit, resulting in 3125 bytes per second. Each OPL write takes up several MIDI bytes. However, USB-MIDI can reach much higher speeds, but I believe there are no guarantees. I did find that while playing very frantic OPL3 music it started to lag a bit in one OS (I don't remember which one). Probably hitting some bandwidth or buffering limit somewhere. I do think that this MIDI approach is viable for the majority of games.

As for the digital sample issue... apart from the high bandwidth that would be required to play a sample, I think there's another big hurdle: timing. Currently we're sending OPL writes over MIDI as they happen with no timing information. If multiple writes happen with small (but important) delays between them, these delays will likely get lost in the MIDI buffers. Even if the bandwidth is sufficient, we probably need a new approach to support playing a sample.

**Comment by vincentbernat** [comment]
With the OPL2LPT and a patch for DOSBox of my own, I get correct results with Another World at 10 kHz (similar to an emulated Sound Blaster), but I have yet to find a true recording of this intro with a real AdLib. There is no buffering with the parallel port, but I don't think it matter much as long as the delays introduced are fixed. I have also a similar working patch for an Arduino OPL2 board over serial port over USB. Due to the limited bandwidth (115200, didn't succeed in using higher serial speeds), I am only able to use the 5 kHz version. Again, no special tentative to get the timing right. This is somewhat similar to what the OPL2LPT is producing at the same rate but the quality drop is quite significant. But again, dunno what a real AdLib would sound. The other difficulty is that DOSBox emulation is not faithful either, even when using Nuked OPL3 and 49.7 kHz sampling. This is easy to spot as the sound of the car "squiding" is not reproduced correctly. If you try and want to compare with the 5 kHz version I get with this Arduino board, have a look at the end of the second video here: https://vincent.bernat.im/en/blog/2018-opl2-audio-board.

I had some plan to use raw HID commands instead to get higher bandwidths, but if USB-MIDI is able to sustain higher bandwidths as well, this would be simpler to implement and would work for Windows users too.

**Comment by waltervn** [comment]
I'll try Another World and see what happens with the OPL3-FPGA.

I have a SoundBlaster with a CT1747 in a retro PC, I could make a recording with that if it would be helpful. That's the closest I have to AdLib.

**Comment by waltervn** [comment]
OK, I did a quick test. Unfortunately, it sounds pretty bad at both 5 and 10 kHz. I do hear a continuous stream of digital sound, but it's very low in volume and muffled. I guess the next step would be to verify that OPL3-FPGA plays this properly by using imfplay, before blaming this on MIDI.

BTW, if you ever want to have a chat about any of this, you can find me (waltervn) on freenode's #scummvm channel.

**Comment by vincentbernat** [comment]
When I have some time to work on that, I'll check whatever high speed serial port, MIDI or raw HID provides the better bandwidth, lowest latency and lowest jitter. What is nice with MIDI is that we could abstract the support of all these OPL2/OPL3 boards by using MIDI and support them more easily in DOSBox and ScummVM. For USB based devices, people could use this support directly. For stuff like OPL2LPT, they would still need some soft MIDI driver.

**Comment by waltervn** [comment]
Right. I went for MIDI because dosbox and scummvm already have MIDI backends for multiple operating systems. On top of that, using MIDI might be the most user-friendly way to make this all work. I'm wondering about something though, do the Arduino and/or Pi support USB device mode? On the ZYBO we're using the Linux MIDI USB gadget to make this work.

I'm going to keep this ticket open until I know what's going on with Another World.

**Comment by vincentbernat** [comment]
Some Arduino have native USB capabilities and can behave like a MIDI USB. Some others have a separate ATmega16u2 to drive the USB port and it can be programmed with alternate sketches to behave as any USB device.

The other solution would be to use a software implement of USB, like V-USB and use two digital pins. But it seems it can only handle USB 1.1 without additional hardware.
